server/schedule/placement/fit.go

```diff
@@ -115,48 +114,134 @@ func compareRuleFit(a, b *RuleFit) int {
+		rules:   rules,
 	}
-	return &regionFit
 }
 
-func fitRule(peers []*fitPeer, rule *Rule) *RuleFit {
-	// Ignore peers that does not match label constraints, and that cannot be
-	// transformed to expected role type.
-	peers = filterPeersBy(peers,
-		func(p *fitPeer) bool { return MatchLabelConstraints(p.store, rule.LabelConstraints) },
-		func(p *fitPeer) bool { return p.matchRoleLoose(rule.Role) })
+func (w *fitWorker) run() {
+	w.fitRule(0)
+	w.updateOrphanPeers(0) // All peers go to orphanList when RuleList is empty.
+}
+
+// Pick the most suitable peer combination for the rule.
+// Index specifies the position of the rule.
+// returns true if it replaces `bestFit` with a better alternative.
+func (w *fitWorker) fitRule(index int) bool {
+	if index >= len(w.rules) {
+		return false
+	}
+	// Only consider stores:
+	// 1. Match label constraints
+	// 2. Role match, or can match after transformed.
+	// 3. Not selected by other rules.
+	var candidates []*fitPeer
+	for _, p := range w.peers {
+		if MatchLabelConstraints(p.store, w.rules[index].LabelConstraints) &&
+			p.matchRoleLoose(w.rules[index].Role) &&
+			!p.selected {
+			candidates = append(candidates, p)
+		}
+	}
 
-	if len(peers) <= rule.Count {
-		return newRuleFit(rule, peers)
+	count := w.rules[index].Count
+	if len(candidates) < count {
+		count = len(candidates)
 	}
+	return w.enumPeers(candidates, nil, index, count)
+}
 
-	// TODO: brute force can be improved.
-	var best *RuleFit
-	iterPeers(peers, rule.Count, func(candidates []*fitPeer) {
-		rf := newRuleFit(rule, candidates)
-		if best == nil || compareRuleFit(rf, best) > 0 {
-			best = rf
+// Recursively traverses all feasible peer combinations.
+// For each combination, call `compareBest` to determine whether it is better
+// than the existing option.
+// Returns true if it replaces `bestFit` with a better alternative.
+func (w *fitWorker) enumPeers(candidates, selected []*fitPeer, index int, count int) bool {
+	if len(selected) == count {
+		// We collect enough peers. End recursive.
+		return w.compareBest(selected, index)
+	}
+
+	var better bool
+	for i, p := range candidates {
+		p.selected = true
+		better = w.enumPeers(candidates[i+1:], append(selected, p), index, count) || better
+		p.selected = false
+	}
+	return better
+}
+
+// compareBest checks if the selected peers is better then previous best.
+// Returns true if it replaces `bestFit` with a better alternative.
+func (w *fitWorker) compareBest(selected []*fitPeer, index int) bool {
+	rf := newRuleFit(w.rules[index], selected)
+	cmp := 1
+	if best := w.bestFit.RuleFits[index]; best != nil {
+		cmp = compareRuleFit(rf, best)
+	}
+
+	switch cmp {
+	case 1:
+		w.bestFit.RuleFits[index] = rf
+		// Reset previous result after position index.
+		for i := index + 1; i < len(w.rules); i++ {
+			w.bestFit.RuleFits[i] = nil
+		}
+		w.fitRule(index + 1)
+		w.updateOrphanPeers(index + 1)
+		return true
+	case 0:
+		if w.fitRule(index + 1) {
+			w.bestFit.RuleFits[index] = rf
```

Comment:
Why do we need to update the `w.bestFit.RuleFits[index]` in the condition of `index + 1` has a better rule to fit?

Reply:
Oh I see…When two rules have the same effect, the one which can make the next index have a better rule will become the current best one.

Reply:
true